File: scripts/publication_provider.py
"""Pull-request transport. Git transport and authority live in pending_publication.

The provider must return a complete current list for the exact head/base scope or
raise UnknownOutcome. It must never turn failed reads into an empty list. Tests
use an explicitly identified provider fixture, never a live hosted repository.
"""
import json
import re
import subprocess
from urllib.parse import urlparse


class UnknownOutcome(RuntimeError):
    """The remote result cannot be established; reconcile before another write."""


class ProviderConflict(ValueError):
    pass


class GitHubProvider:
    """GitHub's authenticated CLI; no permission is inferred from authentication."""
    def __init__(self, repository):
        match = re.fullmatch(r'git@([^:]+):([^/]+/[^/]+?)(?:\.git)?', repository)
        if match:
            host, name = match.groups()
        else:
            parsed = urlparse(repository)
            if parsed.scheme not in ('https', 'ssh') or parsed.password or parsed.query or parsed.fragment:
                raise ProviderConflict('configure a supported GitHub repository URL')
            host, name = parsed.hostname, parsed.path.strip('/').removesuffix('.git')
        if not host or not re.fullmatch(r'[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+', name):
            raise ProviderConflict('repository must identify one GitHub owner and repository')
        self.host, self.name = host, name

    def _api(self, method, endpoint, fields=None):
        args = ['gh', 'api', '--hostname', self.host, '--method', method, endpoint]
        if fields is not None:
            args.extend(['--input', '-'])
        try:
            result = subprocess.run(args, input=json.dumps(fields).encode() if fields is not None else None,
                                    stdout=subprocess.PIPE, stderr=subprocess.PIPE, timeout=30)
        except (OSError, subprocess.TimeoutExpired) as error:
            raise UnknownOutcome('provider is unavailable; remote outcome is unknown') from error
        if result.returncode:
            raise UnknownOutcome('provider request failed; reconcile its outcome before retrying')
        try:
            return json.loads(result.stdout)
        except (ValueError, UnicodeError) as error:
            raise UnknownOutcome('provider returned an unreadable response') from error

    def list(self, scope):
        # Pagination is explicit so absence means a completed read, not page one.
        result, page = [], 1
        while True:
            rows = self._api('GET', 'repos/' + self.name + '/pulls?state=all&per_page=100&page=' + str(page))
            if not isinstance(rows, list):
                raise UnknownOutcome('provider did not return a complete pull-request list')
            for row in rows:
                if row['head']['ref'] == scope['branch'] and row['base']['ref'] == scope['target'] \
                        and (row['head'].get('repo') or {}).get('full_name') == self.name \
                        and (row['base'].get('repo') or {}).get('full_name') == self.name:
                    result.append(self._row(row))
            if len(rows) < 100:
                return result
            page += 1
            if page > 100:
                raise UnknownOutcome('provider listing exceeded its bounded pagination limit')

    @staticmethod
    def _row(row):
        return {'id': str(row['number']), 'state': 'merged' if row.get('merged_at') else row['state'],
                'url': row['html_url'], 'head': row['head']['sha'], 'body': row.get('body') or ''}

    def create(self, scope, *, title, body, request_id):
        # GitHub has no idempotency key for this endpoint. The durable caller retains
        # request_id and never repeats an uncertain create without reconciliation.
        return self._row(self._api('POST', 'repos/' + self.name + '/pulls',
                                  {'head': scope['branch'], 'base': scope['target'], 'title': title, 'body': body}))

    def update(self, scope, pr_id, *, title, body):
        return self._row(self._api('PATCH', 'repos/' + self.name + '/pulls/' + str(int(pr_id)),
                                  {'title': title, 'body': body}))

File: native/src/publication_provider.rs
//! Pull-request transport. Git transport and authority remain in
//! `pending_publication`; this adapter never infers permission from login state.
use crate::{Error, Result, require};
use regex::Regex;
use serde_json::{Value, json};
use std::{process::Command, sync::LazyLock, time::Duration};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct PullRequest {
    pub id: String,
    pub state: String,
    pub url: String,
    pub head: String,
    pub body: String,
}

pub trait Provider {
    fn list(&mut self, scope: &Value) -> Result<Vec<PullRequest>>;
    fn create(
        &mut self,
        scope: &Value,
        title: &str,
        body: &str,
        request_id: &str,
    ) -> Result<PullRequest>;
    fn update(&mut self, scope: &Value, id: &str, title: &str, body: &str) -> Result<PullRequest>;
}

pub struct GitHubProvider {
    repository: String,
}

impl GitHubProvider {
    pub fn new(repository: &str) -> Self {
        Self {
            repository: repository.into(),
        }
    }

    fn identity(&self) -> Result<(String, String)> {
        static SSH: LazyLock<Regex> =
            LazyLock::new(|| Regex::new(r"^git@([^:]+):([^/]+/[^/]+?)(?:\.git)?$").unwrap());
        static NAME: LazyLock<Regex> =
            LazyLock::new(|| Regex::new(r"^[A-Za-z0-9_.-]+/[A-Za-z0-9_.-]+$").unwrap());
        let (host, name) = if let Some(found) = SSH.captures(&self.repository) {
            (found[1].to_owned(), found[2].to_owned())
        } else {
            let rest = self
                .repository
                .strip_prefix("https://")
                .or_else(|| self.repository.strip_prefix("ssh://"))
                .ok_or_else(|| Error("configure a supported GitHub repository URL".into()))?;
            require(
                !rest.contains(['?', '#', '@']),
                "configure a supported GitHub repository URL",
            )?;
            let (host, path) = rest.split_once('/').ok_or_else(|| {
                Error("repository must identify one GitHub owner and repository".into())
            })?;
            (
                host.to_owned(),
                path.trim_matches('/').trim_end_matches(".git").to_owned(),
            )
        };
        require(
            !host.is_empty() && NAME.is_match(&name),
            "repository must identify one GitHub owner and repository",
        )?;
        Ok((host, name))
    }

    fn api(&self, method: &str, endpoint: &str, fields: Option<&Value>) -> Result<Value> {
        let (host, _) = self.identity()?;
        let mut command = Command::new("gh");
        command.args(["api", "--hostname", &host, "--method", method, endpoint]);
        let payload = if let Some(fields) = fields {
            command.args(["--input", "-"]);
            serde_json::to_vec(fields)?
        } else {
            vec![]
        };
        let raw = crate::reasoning_runtime::run_command_bounded(
            &mut command,
            payload,
            Duration::from_secs(30),
            16 * 1024 * 1024,
        )
        .map_err(|error| {
            if error.0 == "native reasoning process failed" {
                Error("provider request failed; reconcile its outcome before retrying".into())
            } else {
                Error("provider is unavailable; remote outcome is unknown".into())
            }
        })?;
        crate::json_ingress::parse_slice(&raw, crate::json_ingress::DuplicateKeys::Reject)
            .map_err(|_| Error("provider returned an unreadable response".into()))
    }

    fn row(value: &Value) -> Result<PullRequest> {
        let state = if value.get("merged_at").is_some_and(|value| !value.is_null()) {
            "merged"
        } else {
            value["state"]
                .as_str()
                .ok_or_else(|| Error("provider returned an incomplete publication state".into()))?
        };
        Ok(PullRequest {
            id: value["number"]
                .as_i64()
                .map(|value| value.to_string())
                .or_else(|| value["id"].as_str().map(str::to_owned))
                .ok_or_else(|| Error("provider returned an incomplete publication state".into()))?,
            state: state.into(),
            url: value["html_url"].as_str().unwrap_or_default().into(),
            head: value["head"]["sha"].as_str().unwrap_or_default().into(),
            body: value["body"].as_str().unwrap_or_default().into(),
        })
    }
}

impl Provider for GitHubProvider {
    fn list(&mut self, scope: &Value) -> Result<Vec<PullRequest>> {
        let (_, name) = self.identity()?;
        let branch = scope["branch"].as_str().unwrap_or_default();
        let target = scope["target"].as_str().unwrap_or_default();
        let mut result = vec![];
        for page in 1..=100 {
            let endpoint = format!("repos/{}/pulls?state=all&per_page=100&page={page}", name);
            let rows = self.api("GET", &endpoint, None)?;
            let rows = rows.as_array().ok_or_else(|| {
                Error("provider did not return a complete pull-request list".into())
            })?;
            for row in rows {
                if row["head"]["ref"] == branch
                    && row["base"]["ref"] == target
                    && row["head"]["repo"]["full_name"] == name
                    && row["base"]["repo"]["full_name"] == name
                {
                    result.push(Self::row(row)?);
                }
            }
            if rows.len() < 100 {
                return Ok(result);
            }
        }
        Err(Error(
            "provider listing exceeded its bounded pagination limit".into(),
        ))
    }

    fn create(
        &mut self,
        scope: &Value,
        title: &str,
        body: &str,
        _request_id: &str,
    ) -> Result<PullRequest> {
        let (_, name) = self.identity()?;
        Self::row(&self.api(
            "POST",
            &format!("repos/{name}/pulls"),
            Some(&json!({"head":scope["branch"],"base":scope["target"],"title":title,"body":body})),
        )?)
    }

    fn update(&mut self, _scope: &Value, id: &str, title: &str, body: &str) -> Result<PullRequest> {
        let (_, name) = self.identity()?;
        require(
            id.bytes().all(|byte| byte.is_ascii_digit()),
            "invalid pull request ID",
        )?;
        Self::row(&self.api(
            "PATCH",
            &format!("repos/{name}/pulls/{id}"),
            Some(&json!({"title":title,"body":body})),
        )?)
    }
}
